File: 14-queue/02-implement-stack-using-queue/02-implement-stack-using-queue.ts
class StackUsingQueue {
  q1 = []
  q2 = []
  constructor() {}

  push(val: number) {
    this.q2 = this.q1
    this.q1 = []
    this.q1.push(val)
    while (this.q2.length > 0) {
      this.q1.push(this.q2.shift())
    }
  }

  pop() {
    return this.q1 ? this.q1.shift() : null
  }

  peek() {
    return this.q1 ? this.q1[0] : null
  }

  size() {
    return this.q1.length
  }

  print() {
    const res = []
    for (let i = 0; i < this.q1.length; i++) {
      res.push(this.q1[i])
    }
    console.log(`[${res.join(',')}]`)
  }
}

const stack = new StackUsingQueue()
stack.push(10)
stack.push(20)
stack.push(30)

stack.print()
console.log(stack.pop())
console.log(stack.pop())
stack.print()

File: 14-queue/02-implement-stack-using-queue/02-implement-stack-using-queue.py
from collections import deque


class StackUsingQueue:
  def __init__(self):
    self.q1, self.q2 = deque(), deque()

  def push(self, val):
    self.q2, self.q1 = self.q1, self.q2
    self.q1.append(val)
    while self.q2:
      self.q1.append(self.q2.popleft())

  def pop(self):
    return self.q1.popleft() if self.q1 else None

  def peek(self):
    return self.q1[0] if self.q1 else None

  def size(self):
    return len(self.q1)

  def __str__(self):
    res = []
    for item in self.q1:
      res.append(str(item))
    return ",".join(res)


stack = StackUsingQueue()
stack.push(10)
stack.push(20)
stack.push(30)
print(stack)
print(stack.pop())
print(stack)
print(stack.peek())
print(stack.size())
